crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -306,3 +295,111 @@ pub(crate) fn filter_contained(edits: Vec<Edit>) -> Vec<Edit> {
     }
     filtered
 }
+
+#[derive(Copy, PartialEq, Clone)]
+enum QuoteAnnotatorState {
+    Literal,
+    AnnotatedFirst,
+    AnnotatedRest,
+    Other,
+}
+
+pub(crate) struct QuoteAnnotator<'a> {
+    stylist: &'a Stylist<'a>,
+    semantic: &'a SemanticModel<'a>,
+    state: Vec<QuoteAnnotatorState>,
+    annotation: String,
+}
+
+impl<'a> QuoteAnnotator<'a> {
+    fn new(semantic: &'a SemanticModel<'a>, stylist: &'a Stylist<'a>) -> Self {
+        Self {
+            stylist,
+            semantic,
+            state: Vec::new(),
+            annotation: String::new(),
+        }
+    }
+
+    fn into_annotation(self) -> String {
+        self.annotation
+    }
+}
+
+impl<'a> source_order::SourceOrderVisitor<'a> for QuoteAnnotator<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let generator = Generator::from(self.stylist);
+
+        match expr {
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                let value_expr = generator.expr(value);
+                self.annotation.push_str(&value_expr);
+                self.annotation.push('[');
+
+                if let Some(qualified_name) = self.semantic.resolve_qualified_name(value) {
+                    if self
+                        .semantic
+                        .match_typing_qualified_name(&qualified_name, "Literal")
+                    {
+                        self.state.push(QuoteAnnotatorState::Literal);
+                    } else if self
+                        .semantic
+                        .match_typing_qualified_name(&qualified_name, "Annotated")
+                    {
+                        self.state.push(QuoteAnnotatorState::AnnotatedFirst);
+                    } else {
+                        self.state.push(QuoteAnnotatorState::Other);
+                    }
+                }
+
+                self.visit_expr(slice);
+                self.state.pop();
+                self.annotation.push(']');
+            }
+            Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                let Some((first, remaining)) = elts.split_first() else {
+                    return;
+                };
+                self.visit_expr(first);
+                if let Some(last) = self.state.last_mut() {
+                    if *last == QuoteAnnotatorState::AnnotatedFirst {
+                        *last = QuoteAnnotatorState::AnnotatedRest;
+                    }
+                }
+                for expr in remaining {
+                    self.annotation.push_str(", ");
+                    self.visit_expr(expr);
+                }
+                self.state.pop();
+            }
+            Expr::BinOp(ast::ExprBinOp {
+                left, op, right, ..
+            }) => {
+                debug_assert_eq!(*op, ast::Operator::BitOr);
+                self.visit_expr(left);
+                self.annotation.push_str(" | ");
+                self.visit_expr(right);
+            }
+            _ => {
```

Comment:
There was a branch for `Expr::BoolOp` although I don't think that's required. Can you confirm this? @Glyphack

Reply:
You are correct. I was not very familiar with syntax in type annotation but I don't see anything with a boolean operation.